### ISSUE TITLE
service.running fails on sle11 sp3 and sp4

### ISSUE THREAD

**Issue by tampakrap** [saltstack/salt#31617]
### Description of Issue/Question

service.running state fails on SLE11 SP3 and SP4 machines. It happened in the past as well, where the SLE11 machines were falsely using systemd, see  https://github.com/saltstack/salt/commit/baf238f270172654d4bb954f3fc21590fda98b47

salt minion log from the SLE11SP3 box:

```
2016-03-02 16:15:03,851 [salt.loader                                            ][ERROR   ][15417] Exception raised when processing __virtual__ function for rh_service. Module will not be loaded local variable 'osrelease' referenced before assignment
2016-03-02 16:15:03,851 [salt.loader                                            ][WARNING ][15417] salt.loaded.int.module.rh_service.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'rh_service', please fix this.
2016-03-02 16:15:04,708 [salt.state                                                  ][ERROR   ][15417] An exception occurred in this state: Traceback (most recent call last):
  File "/usr/lib64/python2.6/site-packages/salt/state.py", line 1624, in call
    **cdata['kwargs'])
  File "/usr/lib64/python2.6/site-packages/salt/loader.py", line 1491, in wrapper
    return f(*args, **kwargs)
  File "/usr/lib64/python2.6/site-packages/salt/states/service.py", line 296, in running
    before_toggle_enable_status = __salt__['service.enabled'](name)
  File "/usr/lib64/python2.6/site-packages/salt/loader.py", line 900, in __getitem__
    func = super(LazyLoader, self).__getitem__(item)
  File "/usr/lib64/python2.6/site-packages/salt/utils/lazy.py", line 93, in __getitem__
    raise KeyError(key)
KeyError: 'service.enabled'

2016-03-02 16:15:05,430 [salt.state                                                  ][ERROR   ][15417] An exception occurred in this state: Traceback (most recent call last):
  File "/usr/lib64/python2.6/site-packages/salt/state.py", line 1624, in call
    **cdata['kwargs'])
  File "/usr/lib64/python2.6/site-packages/salt/loader.py", line 1491, in wrapper
    return f(*args, **kwargs)
  File "/usr/lib64/python2.6/site-packages/salt/states/service.py", line 296, in running
    before_toggle_enable_status = __salt__['service.enabled'](name)
  File "/usr/lib64/python2.6/site-packages/salt/loader.py", line 900, in __getitem__
    func = super(LazyLoader, self).__getitem__(item)
  File "/usr/lib64/python2.6/site-packages/salt/utils/lazy.py", line 93, in __getitem__
    raise KeyError(key)
KeyError: 'service.enabled'
```
salt minion log from the SLE11SP4 box:
```
2016-03-02 15:18:44,759 [salt.state                                                  ][ERROR   ][4430] An exception occurred in this state: Traceback (most recent call last):
  File "/usr/lib64/python2.6/site-packages/salt/state.py", line 1624, in call
    **cdata['kwargs'])
  File "/usr/lib64/python2.6/site-packages/salt/loader.py", line 1491, in wrapper
    return f(*args, **kwargs)
  File "/usr/lib64/python2.6/site-packages/salt/states/service.py", line 296, in running
    before_toggle_enable_status = __salt__['service.enabled'](name)
  File "/usr/lib64/python2.6/site-packages/salt/loader.py", line 900, in __getitem__
    func = super(LazyLoader, self).__getitem__(item)
  File "/usr/lib64/python2.6/site-packages/salt/utils/lazy.py", line 93, in __getitem__
    raise KeyError(key)
KeyError: 'service.enabled'

2016-03-02 15:18:45,371 [salt.state                                                  ][ERROR   ][4430] An exception occurred in this state: Traceback (most recent call last):
  File "/usr/lib64/python2.6/site-packages/salt/state.py", line 1624, in call
    **cdata['kwargs'])
  File "/usr/lib64/python2.6/site-packages/salt/loader.py", line 1491, in wrapper
    return f(*args, **kwargs)
  File "/usr/lib64/python2.6/site-packages/salt/states/service.py", line 296, in running
    before_toggle_enable_status = __salt__['service.enabled'](name)
  File "/usr/lib64/python2.6/site-packages/salt/loader.py", line 900, in __getitem__
    func = super(LazyLoader, self).__getitem__(item)
  File "/usr/lib64/python2.6/site-packages/salt/utils/lazy.py", line 93, in __getitem__
    raise KeyError(key)
KeyError: 'service.enabled'
```

### Setup
*Please provide relevant configs and/or SLS files (Be sure to remove sensitive info).*

useful grains:
```
    os:
        SUSE
    os_family:
        Suse
    osarch:
        x86_64
    oscodename:
        n.a
    osfullname:
        SLES
    osrelease:
        11 SP3
    osrelease_info:
        - 11 SP3
```

```
    os:
        SUSE
    os_family:
        Suse
    osarch:
        x86_64
    oscodename:
        SUSE Linux Enterprise Server 11 SP4
    osfullname:
        SLES
    osrelease:
        11.4
    osrelease_info:
        - 11
        - 4
```
### Versions Report
*Provided by running* ``salt --versions-report``
```
# salt-call --versions-report
Salt Version:
           Salt: 2015.8.7
 
Dependency Versions:
         Jinja2: 2.6
       M2Crypto: 0.21.1
           Mako: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.7.0
         Python: 2.6.9 (unknown, Apr  7 2015, 08:28:12)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.2
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
        libgit2: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: Not Installed
        timelib: Not Installed
 
System Versions:
           dist: SuSE 11 x86_64
        machine: x86_64
        release: 3.0.101-68-default
         system: SUSE Linux Enterprise Server  11 x86_64
```

**Comment by darix** [comment]
The real bug is ... the service state assumes that all provider implement a enabled method. Which is not true for modules/service.py. I have a local patch that makes sles11 use the modules/rh_service.py again.

**Comment by darix** [comment]
potential fix here https://github.com/openSUSE/salt/compare/openSUSE-2015.8.7...darix:fix-services-sles

**Comment by jfindlay** [comment]
@darix, you're welcome to submit your changes to 2015.8.

**Comment by Ch3LL** [comment]
I am also ran into this issue with the following test case:

using salt-ssh
salt-version: 2015.8.7

/srv/salt/test-state.sls:

```
disable_services:
  service.dead:
    - names:
      - salt-master
      - salt-minion
```

`salt-ssh 'sles11sp4minion' state.sls test-state`

```
ch3ll-sles11:
----------
          ID: disable_services
    Function: service.dead
        Name: salt-minion
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/tmp/.ec2-user_881c5e_salt/salt/state.py", line 1602, in call
                  **cdata['kwargs'])
                File "/tmp/.ec2-user_881c5e_salt/salt/loader.py", line 1491, in wrapper
                  return f(*args, **kwargs)
                File "/tmp/.ec2-user_881c5e_salt/salt/states/service.py", line 379, in dead
                  before_toggle_enable_status = __salt__['service.enabled'](name)
                File "/tmp/.ec2-user_881c5e_salt/salt/loader.py", line 900, in __getitem__
                  func = super(LazyLoader, self).__getitem__(item)
                File "/tmp/.ec2-user_881c5e_salt/salt/utils/lazy.py", line 93, in __getitem__
                  raise KeyError(key)
              KeyError: 'service.enabled'
     Started: 18:20:28.541897
    Duration: 2137.492 ms
     Changes:   
----------
          ID: disable_services
    Function: service.dead
        Name: salt-master
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/tmp/.ec2-user_881c5e_salt/salt/state.py", line 1602, in call
                  **cdata['kwargs'])
                File "/tmp/.ec2-user_881c5e_salt/salt/loader.py", line 1491, in wrapper
                  return f(*args, **kwargs)
                File "/tmp/.ec2-user_881c5e_salt/salt/states/service.py", line 379, in dead
                  before_toggle_enable_status = __salt__['service.enabled'](name)
                File "/tmp/.ec2-user_881c5e_salt/salt/loader.py", line 900, in __getitem__
                  func = super(LazyLoader, self).__getitem__(item)
                File "/tmp/.ec2-user_881c5e_salt/salt/utils/lazy.py", line 93, in __getitem__
                  raise KeyError(key)
             KeyError: 'service.enabled'
     Started: 18:20:30.679583
    Duration: 42.359 ms
     Changes:   

Summary for ch3ll-sles11
------------
Succeeded: 0
Failed:    2
------------
Total states run:     2
```


This was working in 2015.8.3 so I did a git bisect. Here is the git bisect:

```
9975508f865ea1152f200b8edf00f9da5d2dda8d is the first bad commit
commit 9975508f865ea1152f200b8edf00f9da5d2dda8d
Author: Justin Findlay <jfindlay@gmail.com>
Date:   Fri Dec 4 14:32:35 2015 -0700

    modules.rh_service.__virtual__: handle SUSE osrelease as num

:040000 040000 2882f370419fcb020886a8d5c088e5bbef604022 0bf86f7cd9ed77c1185d149af9d10c3f63f3a882 M      salt
```

**Comment by darix** [comment]
didnt test develop if it is also affected.

**Comment by darix** [comment]
also wondering if salt/modules/service.py should be enhanced to be usable with service state

**Comment by jfindlay** [comment]
@darix, all modules that provide `service` including `salt/modules/service.py` should be providing a common core API to the service state.

**Comment by rallytime** [comment]
@tampakrap and @Ch3LL Can you give the fix in #31629 a try and confirm if this is fixed?

**Comment by tampakrap** [comment]
confirmed thanks!
